File: src/crafting/RecipeRegistry.ts
import { Block } from '../world/blocks';
import { Item } from '../player/items';
import type { RecipeDefinition } from './RecipeDefinition';

export class RecipeRegistry {
  private static instance: RecipeRegistry | null = null;
  private readonly recipes = new Map<string, RecipeDefinition>();

  static get(): RecipeRegistry {
    if (!this.instance) {
      this.instance = new RecipeRegistry();
      this.instance.registerDefaults();
    }
    return this.instance;
  }

  register(recipe: RecipeDefinition): void {
    this.recipes.set(recipe.id, Object.freeze({ ...recipe }));
  }

  get(id: string): RecipeDefinition | undefined {
    return this.recipes.get(id);
  }

  has(id: string): boolean {
    return this.recipes.has(id);
  }

  getAll(): RecipeDefinition[] {
    return Array.from(this.recipes.values());
  }

  private registerDefaults(): void {
    // ── BASIC CRAFTING ──
    this.register({
      id: 'planks',
      name: 'Oak Planks',
      type: 'SHAPELESS',
      ingredients: [Block.Wood],
      result: { id: Block.Planks, count: 4 },
      hint: '1 Oak Log → 4 Planks',
    });

    this.register({
      id: 'sticks',
      name: 'Sticks',
      type: 'SHAPED',
      ingredients: [
        Block.Planks, 0, 0,
        Block.Planks, 0, 0,
        0, 0, 0,
      ],
      result: { id: Item.Stick, count: 4 },
      hint: '2 Planks (vertical) → 4 Sticks',
    });

    this.register({
      id: 'crafting_table',
      name: 'Crafting Table',
      type: 'SHAPED',
      ingredients: [
        Block.Planks, Block.Planks, 0,
        Block.Planks, Block.Planks, 0,
        0, 0, 0,
      ],
      result: { id: Block.CraftingTable, count: 1 },
      hint: '4 Planks (2×2) → Crafting Table',
    });

    this.register({
      id: 'torches',
      name: 'Torches',
      type: 'SHAPED',
      ingredients: [
        Item.Coal, 0, 0,
        Item.Stick, 0, 0,
        0, 0, 0,
      ],
      result: { id: Block.Torch, count: 4 },
      hint: 'Coal over Stick → 4 Torches',
    });

    // ── WOODEN TOOLS (3x3) ──
    this.register({
      id: 'wooden_pickaxe',
      name: 'Wooden Pickaxe',
      type: 'SHAPED',
      ingredients: [
        Block.Planks, Block.Planks, Block.Planks,
        0, Item.Stick, 0,
        0, Item.Stick, 0,
      ],
      result: { id: Item.WoodenPickaxe, count: 1, durability: 60, maxDurability: 60 },
      hint: '3 Planks top row + 2 Sticks center → Wooden Pickaxe',
      gridRequired: '3x3',
    });

    this.register({
      id: 'wooden_axe',
      name: 'Wooden Axe',
      type: 'SHAPED',
      ingredients: [
        Block.Planks, Block.Planks, 0,
        Block.Planks, Item.Stick, 0,
        0, Item.Stick, 0,
      ],
      result: { id: Item.WoodenAxe, count: 1, durability: 60, maxDurability: 60 },
      hint: '3 Planks corner + 2 Sticks → Wooden Axe',
      gridRequired: '3x3',
    });

    this.register({
      id: 'wooden_sword',
      name: 'Wooden Sword',
      type: 'SHAPED',
      ingredients: [
        Block.Planks, 0, 0,
        Block.Planks, 0, 0,
        Item.Stick, 0, 0,
      ],
      result: { id: Item.WoodenSword, count: 1, durability: 60, maxDurability: 60 },
      hint: '2 Planks over 1 Stick → Wooden Sword',
    });

    // ── STONE TOOLS (3x3) ──
    this.register({
      id: 'stone_pickaxe',
      name: 'Stone Pickaxe',
      type: 'SHAPED',
      ingredients: [
        Block.Cobblestone, Block.Cobblestone, Block.Cobblestone,
        0, Item.Stick, 0,
        0, Item.Stick, 0,
      ],
      result: { id: Item.StonePickaxe, count: 1, durability: 132, maxDurability: 132 },
      hint: '3 Cobblestone top row + 2 Sticks center → Stone Pickaxe',
      gridRequired: '3x3',
    });

    this.register({
      id: 'stone_axe',
      name: 'Stone Axe',
      type: 'SHAPED',
      ingredients: [
        Block.Cobblestone, Block.Cobblestone, 0,
        Block.Cobblestone, Item.Stick, 0,
        0, Item.Stick, 0,
      ],
      result: { id: Item.StoneAxe, count: 1, durability: 132, maxDurability: 132 },
      hint: '3 Cobblestone corner + 2 Sticks → Stone Axe',
      gridRequired: '3x3',
    });

    this.register({
      id: 'stone_sword',
      name: 'Stone Sword',
      type: 'SHAPED',
      ingredients: [
        Block.Cobblestone, 0, 0,
        Block.Cobblestone, 0, 0,
        Item.Stick, 0, 0,
      ],
      result: { id: Item.StoneSword, count: 1, durability: 132, maxDurability: 132 },
      hint: '2 Cobblestone over 1 Stick → Stone Sword',
    });

    // ── IRON TOOLS & ARMOR (3x3) ──
    this.register({
      id: 'iron_pickaxe',
      name: 'Iron Pickaxe',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, Item.IronIngot, Item.IronIngot,
        0, Item.Stick, 0,
        0, Item.Stick, 0,
      ],
      result: { id: Item.IronPickaxe, count: 1, durability: 251, maxDurability: 251 },
      hint: '3 Iron Ingots + 2 Sticks → Iron Pickaxe',
      gridRequired: '3x3',
    });

    this.register({
      id: 'iron_sword',
      name: 'Iron Sword',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, 0, 0,
        Item.IronIngot, 0, 0,
        Item.Stick, 0, 0,
      ],
      result: { id: Item.IronSword, count: 1, durability: 251, maxDurability: 251 },
      hint: '2 Iron Ingots over 1 Stick → Iron Sword',
    });

    this.register({
      id: 'iron_helmet',
      name: 'Iron Helmet',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, Item.IronIngot, Item.IronIngot,
        Item.IronIngot, 0, Item.IronIngot,
        0, 0, 0,
      ],
      result: { id: Item.IronHelmet, count: 1, durability: 165, maxDurability: 165 },
      hint: '5 Iron Ingots in arch → Iron Helmet',
      gridRequired: '3x3',
    });

    this.register({
      id: 'iron_chestplate',
      name: 'Iron Chestplate',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, 0, Item.IronIngot,
        Item.IronIngot, Item.IronIngot, Item.IronIngot,
        Item.IronIngot, Item.IronIngot, Item.IronIngot,
      ],
      result: { id: Item.IronChestplate, count: 1, durability: 240, maxDurability: 240 },
      hint: '8 Iron Ingots in shirt pattern → Iron Chestplate',
      gridRequired: '3x3',
    });

    this.register({
      id: 'iron_leggings',
      name: 'Iron Leggings',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, Item.IronIngot, Item.IronIngot,
        Item.IronIngot, 0, Item.IronIngot,
        Item.IronIngot, 0, Item.IronIngot,
      ],
      result: { id: Item.IronLeggings, count: 1, durability: 225, maxDurability: 225 },
      hint: '7 Iron Ingots in pants pattern → Iron Leggings',
      gridRequired: '3x3',
    });

    this.register({
      id: 'iron_boots',
      name: 'Iron Boots',
      type: 'SHAPED',
      ingredients: [
        Item.IronIngot, 0, Item.IronIngot,
        Item.IronIngot, 0, Item.IronIngot,
        0, 0, 0,
      ],
      result: { id: Item.IronBoots, count: 1, durability: 195, maxDurability: 195 },
      hint: '4 Iron Ingots in 2 columns → Iron Boots',
      gridRequired: '3x3',
    });
  }
}

File: Assets/VYTHERA/Gameplay/Crafting/RecipeRegistry.cs
using System;
using System.Collections.Generic;
using VYTHERA.Gameplay.Inventory;
using VYTHERA.Voxel.Data;

namespace VYTHERA.Gameplay.Crafting
{
    public enum RecipeType
    {
        Shaped,
        Shapeless
    }

    public struct RecipeDefinition
    {
        public string Id;
        public string Name;
        public RecipeType Type;
        public int[] Ingredients; // 3x3 matrix (9 items) or shapeless list
        public ItemStack Result;
        public bool Requires3x3;
    }

    public static class RecipeRegistry
    {
        public static readonly List<RecipeDefinition> Recipes = new List<RecipeDefinition>(32);

        static RecipeRegistry()
        {
            // Planks (1 Wood -> 4 Planks)
            Recipes.Add(new RecipeDefinition
            {
                Id = "planks",
                Name = "Oak Planks",
                Type = RecipeType.Shapeless,
                Ingredients = new int[] { (int)BlockType.Wood },
                Result = new ItemStack((int)BlockType.Planks, 4),
                Requires3x3 = false
            });

            // Sticks (2 Planks vertical -> 4 Sticks)
            Recipes.Add(new RecipeDefinition
            {
                Id = "sticks",
                Name = "Sticks",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    (int)BlockType.Planks, 0, 0,
                    (int)BlockType.Planks, 0, 0,
                    0, 0, 0
                },
                Result = new ItemStack(ItemRegistry.ItemStick, 4),
                Requires3x3 = false
            });

            // Crafting Table (4 Planks 2x2 -> 1 Crafting Table)
            Recipes.Add(new RecipeDefinition
            {
                Id = "crafting_table",
                Name = "Crafting Table",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    (int)BlockType.Planks, (int)BlockType.Planks, 0,
                    (int)BlockType.Planks, (int)BlockType.Planks, 0,
                    0, 0, 0
                },
                Result = new ItemStack((int)BlockType.CraftingTable, 1),
                Requires3x3 = false
            });

            // Torches (1 Coal over 1 Stick -> 4 Torches)
            Recipes.Add(new RecipeDefinition
            {
                Id = "torches",
                Name = "Torches",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    ItemRegistry.ItemCoal, 0, 0,
                    ItemRegistry.ItemStick, 0, 0,
                    0, 0, 0
                },
                Result = new ItemStack((int)BlockType.Torch, 4),
                Requires3x3 = false
            });

            // Wooden Pickaxe (3 Planks top row, 2 Sticks center column)
            Recipes.Add(new RecipeDefinition
            {
                Id = "wooden_pickaxe",
                Name = "Wooden Pickaxe",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    (int)BlockType.Planks, (int)BlockType.Planks, (int)BlockType.Planks,
                    0, ItemRegistry.ItemStick, 0,
                    0, ItemRegistry.ItemStick, 0
                },
                Result = new ItemStack(ItemRegistry.ItemWoodenPickaxe, 1, 60, 60),
                Requires3x3 = true
            });

            // Stone Pickaxe (3 Cobblestone top row, 2 Sticks center column)
            Recipes.Add(new RecipeDefinition
            {
                Id = "stone_pickaxe",
                Name = "Stone Pickaxe",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    (int)BlockType.Cobblestone, (int)BlockType.Cobblestone, (int)BlockType.Cobblestone,
                    0, ItemRegistry.ItemStick, 0,
                    0, ItemRegistry.ItemStick, 0
                },
                Result = new ItemStack(ItemRegistry.ItemStonePickaxe, 1, 132, 132),
                Requires3x3 = true
            });

            // Iron Pickaxe (3 Iron Ingots top row, 2 Sticks center column)
            Recipes.Add(new RecipeDefinition
            {
                Id = "iron_pickaxe",
                Name = "Iron Pickaxe",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot,
                    0, ItemRegistry.ItemStick, 0,
                    0, ItemRegistry.ItemStick, 0
                },
                Result = new ItemStack(ItemRegistry.ItemIronPickaxe, 1, 251, 251),
                Requires3x3 = true
            });

            // Wooden Sword (2 Planks over 1 Stick)
            Recipes.Add(new RecipeDefinition
            {
                Id = "wooden_sword",
                Name = "Wooden Sword",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    (int)BlockType.Planks, 0, 0,
                    (int)BlockType.Planks, 0, 0,
                    ItemRegistry.ItemStick, 0, 0
                },
                Result = new ItemStack(ItemRegistry.ItemWoodenSword, 1, 60, 60),
                Requires3x3 = false
            });

            // Iron Sword (2 Iron Ingots over 1 Stick)
            Recipes.Add(new RecipeDefinition
            {
                Id = "iron_sword",
                Name = "Iron Sword",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    ItemRegistry.ItemIronIngot, 0, 0,
                    ItemRegistry.ItemIronIngot, 0, 0,
                    ItemRegistry.ItemStick, 0, 0
                },
                Result = new ItemStack(ItemRegistry.ItemIronSword, 1, 251, 251),
                Requires3x3 = false
            });

            // Iron Chestplate (8 Iron Ingots shirt pattern)
            Recipes.Add(new RecipeDefinition
            {
                Id = "iron_chestplate",
                Name = "Iron Chestplate",
                Type = RecipeType.Shaped,
                Ingredients = new int[]
                {
                    ItemRegistry.ItemIronIngot, 0, ItemRegistry.ItemIronIngot,
                    ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot,
                    ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot, ItemRegistry.ItemIronIngot
                },
                Result = new ItemStack(ItemRegistry.ItemIronChestplate, 1, 240, 240),
                Requires3x3 = true
            });
        }

        public static ItemStack MatchRecipe(int[] grid, int width, int height)
        {
            if (grid == null) return ItemStack.Empty;

            for (int r = 0; r < Recipes.Count; r++)
            {
                var recipe = Recipes[r];
                if (recipe.Requires3x3 && (width < 3 || height < 3)) continue;

                if (recipe.Type == RecipeType.Shapeless)
                {
                    if (MatchShapeless(grid, recipe.Ingredients)) return recipe.Result;
                }
                else
                {
                    if (MatchShaped(grid, width, height, recipe.Ingredients)) return recipe.Result;
                }
            }

            return ItemStack.Empty;
        }

        private static bool MatchShapeless(int[] grid, int[] ingredients)
        {
            var gridItems = new List<int>();
            for (int i = 0; i < grid.Length; i++)
            {
                if (grid[i] != 0) gridItems.Add(grid[i]);
            }

            if (gridItems.Count != ingredients.Length) return false;

            var ingList = new List<int>(ingredients);
            for (int i = 0; i < gridItems.Count; i++)
            {
                if (!ingList.Remove(gridItems[i])) return false;
            }

            return ingList.Count == 0;
        }

        private static bool MatchShaped(int[] grid, int width, int height, int[] recipe3x3)
        {
            // Pad or test 3x3 matrix directly
            if (width == 3 && height == 3)
            {
                for (int i = 0; i < 9; i++)
                {
                    if (grid[i] != recipe3x3[i]) return false;
                }
                return true;
            }
            return false;
        }
    }
}
